### PR TITLE
Feature/add war files

### PR DESCRIPTION
**Pull request by Frickerson**

### CODE REVIEW

File: Source/WorldActorRegistry/Private/WARWorldActorRegistrationComponent.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
+#include "WorldActorRegistry/Public/WARWorldActorRegistrationComponent.h"
```
Comment (TheEmidee) [inline Source/WorldActorRegistry/Private/WARWorldActorRegistrationComponent.cpp]: Only "WorldActorRegistry/Public/WARWorldActorRegistrationComponent.h"

you never have to have the Public folder in the include path

File: Source/WorldActorRegistry/Private/WARWorldActorRegistrationComponent.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#include "WorldActorRegistry/Public/WARWorldActorRegistrationComponent.h"
+
+#include "WorldActorRegistry/Public/WARWorldActorRegistry.h"
```
Comment (TheEmidee) [inline Source/WorldActorRegistry/Private/WARWorldActorRegistrationComponent.cpp]: same

File: Source/WorldActorRegistry/Private/WARWorldActorRegistrationComponent.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+void UWARWorldActorRegistrationComponent::BeginPlay()
+{
+    Super::BeginPlay();
+
+    if ( !GetOwner()->HasAuthority() )
+    {
+        return;
+    }
+
+    if ( auto * world = GetWorld() )
```
Comment (TheEmidee) [inline Source/WorldActorRegistry/Private/WARWorldActorRegistrationComponent.cpp:21]: I would add an `if ( ensure( GameplayTag.IsValid() ) )`

File: Source/WorldActorRegistry/Private/WARWorldActorRegistry.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    if ( actor == nullptr ||
+         !actor->HasAuthority() ||
+         Registry.Num() == 0 )
+    {
+        return false;
+    }
+
+    auto & existing_actors = Registry.FindOrAdd( actor->GetClass() );
+
+    if ( existing_actors.Find( tag ) )
```
Comment (TheEmidee) [inline Source/WorldActorRegistry/Private/WARWorldActorRegistry.cpp]: I would add an ensure. It's better to warn the user that the registration failed

File: Source/WorldActorRegistry/Private/WARWorldActorRegistry.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    return nullptr;
+}
+
+bool UWARWorldActorRegistry::AddActorToRegistry( AActor * actor, const FGameplayTag tag )
+{
+    if ( actor == nullptr ||
+         !actor->HasAuthority() ||
+         Registry.Num() == 0 )
```
Comment (TheEmidee) [inline Source/WorldActorRegistry/Private/WARWorldActorRegistry.cpp]: why `Registry.Num() == 0` ?

File: Source/WorldActorRegistry/Private/WARWorldActorRegistry.cpp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    existing_actors.Add( tag, actor );
+    return true;
+}
+
+bool UWARWorldActorRegistry::RemoveActorFromRegistry( AActor * actor, const FGameplayTag tag )
+{
+    if ( actor == nullptr ||
+         !actor->HasAuthority() ||
+         Registry.Num() == 0 )
```
Comment (TheEmidee) [inline Source/WorldActorRegistry/Private/WARWorldActorRegistry.cpp:61]: same

File: Source/WorldActorRegistry/Public/WARWorldActorRegistry.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+#include "GameplayTagContainer.h"
+
+#include <CoreMinimal.h>
+#include <Subsystems/WorldSubsystem.h>
+
+#include "WARWorldActorRegistry.generated.h"
+
+UCLASS()
+class WORLDACTORREGISTRY_API UWARWorldActorRegistry : public UWorldSubsystem {
```
Comment (TheEmidee) [inline Source/WorldActorRegistry/Public/WARWorldActorRegistry.h]: CS

File: Source/WorldActorRegistry/Public/WARWorldActorRegistry.h [hunk 8]
```diff
@@ -0,0 +1,37 @@
+
+#pragma once
+
+#include "GameplayTagContainer.h"
```
Comment (TheEmidee) [inline Source/WorldActorRegistry/Public/WARWorldActorRegistry.h]: <>

File: Source/WorldActorRegistry/Public/WARWorldActorRegistry.h [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+#include <Subsystems/WorldSubsystem.h>
+
+#include "WARWorldActorRegistry.generated.h"
+
+UCLASS()
+class WORLDACTORREGISTRY_API UWARWorldActorRegistry : public UWorldSubsystem {
+  GENERATED_BODY()
+
+public:
+  UFUNCTION(BlueprintPure)
```
Comment (TheEmidee) [inline Source/WorldActorRegistry/Public/WARWorldActorRegistry.h]: CS (missing spaces a bit everywhere)

File: Source/WorldActorRegistry/Public/WARWorldActorRegistry.h [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  TArray<AActor *> GetActorsFromClass(UClass *actor_class) const;
+
+  UFUNCTION(BlueprintPure)
+  AActor *GetActorFromClass(UClass *actor_class) const;
+
+  UFUNCTION(BlueprintPure)
+  AActor *GetActorFromClassWithTag(UClass *actor_class,
```
Comment (TheEmidee) [inline Source/WorldActorRegistry/Public/WARWorldActorRegistry.h]: CS

**Review by TheEmidee — CHANGES_REQUESTED** [review]

File: Source/WorldActorRegistry/Private/WARWorldActorRegistrationComponent.cpp (hunk 3 above)
Comment (Frickerson) [reply Source/WorldActorRegistry/Private/WARWorldActorRegistrationComponent.cpp:21]: Done.

File: Source/WorldActorRegistry/Private/WARWorldActorRegistry.cpp (hunk 4 above)
Comment (Frickerson) [reply Source/WorldActorRegistry/Private/WARWorldActorRegistry.cpp]: Done.

File: Source/WorldActorRegistry/Private/WARWorldActorRegistry.cpp (hunk 5 above)
Comment (Frickerson) [reply Source/WorldActorRegistry/Private/WARWorldActorRegistry.cpp]: Removed. Was a stupid mistake.

File: Source/WorldActorRegistry/Private/WARWorldActorRegistry.cpp (hunk 6 above)
Comment (Frickerson) [reply Source/WorldActorRegistry/Private/WARWorldActorRegistry.cpp:61]: If the registry doesn't contain anything, it won't need to try and remove anything.

File: Source/WorldActorRegistry/Public/WARWorldActorRegistry.h (hunk 7 above)
Comment (Frickerson) [reply Source/WorldActorRegistry/Public/WARWorldActorRegistry.h]: Done.

File: Source/WorldActorRegistry/Public/WARWorldActorRegistry.h (hunk 8 above)
Comment (Frickerson) [reply Source/WorldActorRegistry/Public/WARWorldActorRegistry.h]: Done.

File: Source/WorldActorRegistry/Public/WARWorldActorRegistry.h (hunk 9 above)
Comment (Frickerson) [reply Source/WorldActorRegistry/Public/WARWorldActorRegistry.h]: Done.

File: Source/WorldActorRegistry/Public/WARWorldActorRegistry.h (hunk 10 above)
Comment (Frickerson) [reply Source/WorldActorRegistry/Public/WARWorldActorRegistry.h]: Done.

**Review by TheEmidee — APPROVED** [review]

**Review by TheEmidee — CHANGES_REQUESTED** [review]
You still need to fix the includes at the top of the CPP files

File: Source/WorldActorRegistry/Private/WARWorldActorRegistrationComponent.cpp (hunk 2 above)
Comment (Frickerson) [reply Source/WorldActorRegistry/Private/WARWorldActorRegistrationComponent.cpp]: Done.

File: Source/WorldActorRegistry/Private/WARWorldActorRegistrationComponent.cpp (hunk 1 above)
Comment (Frickerson) [reply Source/WorldActorRegistry/Private/WARWorldActorRegistrationComponent.cpp]: Done.

**Review by TheEmidee — APPROVED** [review]

**Pull request merged**